File: Algorithms/基础/哈希/哈希表/判断集合相等或包含/包含/LC76.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"
typedef long long LL;
typedef unsigned long long ULL;
typedef pair<int, int> PII;
const int INF = 0x3f3f3f3f, MOD = 1e9 + 7;

// https://leetcode.cn/problems/minimum-window-substring/

class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> mp;
        for (auto& c : t) mp[c]++;
        int j = 0, cnt = 0, maxlen = s.size() + 1, start;
        for (int i = 0; i < s.size(); ++i) {
            if (--mp[s[i]] >= 0) ++cnt;
            while (cnt == t.size()) {
                if (maxlen > i - j + 1) {
                    maxlen = i - j + 1;
                    start = j;
                }
                if (++mp[s[j]] > 0) cnt--;
                j++;
            }
        }
        return maxlen == s.size() + 1 ? "" : s.substr(start, maxlen);
    }
};

int main() {
    system("pause");
    return 0;
}

File: 数据结构/基础/哈希/哈希表/判断集合相等或包含/包含/lc76.go
package main

// https://leetcode.cn/problems/minimum-window-substring/description/

func minWindow(s string, t string) (res string) {
	cnt := [128]int{}
	for i := range t {
		cnt[t[i]]++
	}
	var sum, j int
	for i := range s {
		if cnt[s[i]]--; cnt[s[i]] >= 0 {
			sum++
		}
		for sum == len(t) {
			if len(res) == 0 || i-j+1 < len(res) {
				res = s[j : i+1]
			}
			c := s[j]
			if cnt[c]++; cnt[c] > 0 {
				sum--
			}
			j++
		}
	}
	return
}
